# Intro to pandas 

<center><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/1920px-Pandas_logo.svg.png" height="200"/></center>

Today we're going to be starting our journey through the pandas library (created by Wes McKinney). The name is derived from the term "panel data", an econometrics term for data sets that include observations over multiple time periods for the same individuals.

When we refer to working in pandas, we're typically talking about working within a **DataFrame**, which is what we'll focus on today. 

## Learning Objectives

At the end of this notebook you will be able to:

- explain why Pandas is important to Data Scientists
- create Pandas DataFrames from lists or a list of dictionaries
- load data from csv-files to DataFrames
- access data stored in DataFrames
- use the common features of a DataFrame


We'll be diving into pandas DataFrames, which are objects that will hold our data, allowing us to interact with it, manipulate it, and eventually throw it into machine learning algorithms (if we want). 

Since a pandas DataFrame is an **object**, this means that we're going to interact with it in much the same way that we interact with all of our other objects in python. Before we get to actually interacting with DataFrames, though, we'll have to get one, and get one with data in it! There's one quick step that we have to do before that... 

## Pandas Import 



```python
import pandas as pd # Standard import. 
```

Here I've shown how we get access to everything in the pandas library - we import it! Also note the python comment `"# Standard import"` out to the right of our import. This was to note that this is the standard way to import the pandas library. 


We should always be sure that if we are importing the entire pandas library, we follow this syntax. It's common practice to use `pd` as the alias, and we tend to follow common practice whenever possible. This makes it easier for others to read our code.

## Getting a DataFrame Object

There are two basic ways that we can get a pandas DataFrame object to work with. The first is by using data that is already in our Python program, in conjunction with the `DataFrame` constructor. The second is by reading in external data through the pandas module (which we've imported and made accessible via `pd`). For reference, here are the [docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) for pandas DataFrames. 

##### Using data already in our Python program

If we are using data that is already in our Python program, then we are going to be passing that data to the `DataFrame` constructor. We typically do this in one of two ways. The first involves passing in a list of dictionaries, whereas the second involves passing in two lists. Let's dive into the first...

In [3]:
import pandas as pd # We haven't actually done this in code yet. 
data_lst = [{'a': 1, 'b': 2, 'c':3}, {'a': 4, 'b':5, 'c':6, 'd':7}]
df = pd.DataFrame(data_lst) 
df

a  b  c    d
0  1  2  3  NaN
1  4  5  6  7.0

DataFrames consist of **rows** and **columns**, where the columns will be denoted and accessed via their names, and the rows will be denoted and access via the indices of the DataFrame. Above, we can look at our columns and see that their names are `a`, `b`, `c`, and `d`. We can similarly look at our rows and see that they are indexed by `0` and `1`. These column names and indices are how we will access this data later. How did the `DataFrame` constructor take our list of dictionaries and put it into the DataFrame in that format, though?

When the pandas DataFrame constructor encounters a list of dictionaries like we gave it, it interprets each dictionary to be a row in the DataFrame. The keys are read as the column names and the values as the values for each column. By default, the DataFrame constructor will assign a column for **every** key that it sees in **any** dictionary in the list of dictionaries. If a particular dictionary in that list doesn't have a value for that key, then it assigns a `NaN` (stands for "not a number") value for that index-column pair. Therefore, when the pandas DataFrame above got the list of dictionaries, it saw `a`, `b`, `c`, and `d` keys, and thus created those columns. It then filled in the values associated with those keys, filling in a `NaN` if it didn't find that key (like it didn't find `d` in the first dictionary in our list). 

In [4]:
data_lst = [{'a': 1}, {'b':5}, {'c': 4}]
df = pd.DataFrame(data_lst)

**What do you expect our DataFrame to hold now?**

In [5]:
df

a    b    c
0  1.0  NaN  NaN
1  NaN  5.0  NaN
2  NaN  NaN  4.0

The second way of creating a DataFrame from data that is already in our Python program is to pass in a list of lists as the `data` argument, and a list of strings as the `columns` argument. The `pd.DataFrame()` constructor will assume that each individual list in the `data` argument is one row (i.e. if you pass in a list of 5 lists, your DataFrame will have 5 rows). Below, we're passing in a list of 2 lists to the `data` parameter, which means that our DataFrame will have two rows.

In [7]:
data_vals = [[1, 2, 3], [4, 5, 6]]
data_cols = ['a', 'b', 'c']
df = pd.DataFrame(data=data_vals, columns=data_cols)
df

a  b  c
0  1  2  3
1  4  5  6

In [10]:
data_col_name = ['a', 'b', 'c']
data_val = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
df = pd.DataFrame(columns=data_col_name, data=data_val)
df

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

It's important to note that this method is not quite as flexible as using a list of dictionaries. When passing in a list of lists via the `data` argument, we have to make sure that the greatest number of elements in any single list corresponds to the number of column names we are passing in via the `columns` argument (no more, no less). For example:

In [ ]:
data_vals = [[1, 2], [4, 5, 6]]
data_cols = ['a', 'b']
df = pd.DataFrame(data=data_vals, columns=data_cols)

We do, however, have the flexibility of passing in some rows with 'missing' data. The key to note here, though, is that the last column name will become what is filled with a `NaN` if a list is missing a column value (it will not be based off a key name like in our list of dictionaries). 

In [12]:
data_vals = [[1, 2], [4, 5, 6]]
data_cols = ['a', 'b', 'c']
df = pd.DataFrame(data=data_vals, columns=data_cols)
df

a  b    c
0  1  2  NaN
1  4  5  6.0

#### Reading External Data


There are many ways that we can read external data into a pandas DataFrame, and they will be called as a function that is available via the `pandas` module. 


As a result of importing the `pandas` module and making it accessible via `pd`, this means that we will call all of these functions via `pd`. Each one of these functions will **return** back to us a pandas DataFrame object, populated with the external data that we read in. 

The [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html) will show you all of the ways that you could load external data into a DataFrame. Basically, there is a way to load in data stored in any format (CSV, JSON, SQL, Excel, HTML). All of these take some form of a `read_{data_type}` function, which means that we will call them as `pd.read_{data_type}`.

So, if we wanted to load data in from a CSV, we would simply use:

```python
df = pd.read_csv('my_data.csv')
```

Note: This assumes that we have the column names in the first row of your .csv.

If we don't have the column names in the first row of our .csv, we could read in the .csv with the following:

```python
df = pd.read_csv('my_data.csv', header=None)
```

Note: This by default assigns numbers as column names (starting with 0).

If we wanted to assign the column names as we read it in, we can pass in an additional `names` argument, where this `names` argument holds a list of the names we want to assign to the columns.

```python
df = pd.read_csv('my_data.csv', header=None, names=['col1', 'col2', ...., 'col12'])
```  

The default assumed column separator is a comma. In case your csv file uses a different one, you can use the `sep` parameter (or `delimiter` which is an alias of `sep`.) It is a necessary skill to realise that there may already be a solution to such problems. In fact, `sep` is described in the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html).

```python
df = pd.read_csv('my_data.csv', ..., sep=';')
```

## Looking at Our Data

Our data is available in the `data` folder inside this repo. It's called `abalone.csv`. We will first read in the data, and then discuss what it's all about.

In [13]:
df = pd.read_csv("data/abalone.csv")

Given that our `DataFrame` is an object, we can imagine that it will have associated attributes and methods. There are a couple of each available to us to get a general sense of our data. We have two attributes that we will frequently use on our DataFrame - these will allow us to look at the shape of our data and the column names. We have four methods that are available on our DataFrame for getting a general sense of our data: `info()`, `describe()`, `head()`, and `tail()`. Let's take a look at what these do. 

In [14]:
df.head()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010  0.150   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485  0.070   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415  0.210   
3   M   0.440     0.365   0.125        0.5160          0.2155   0.1140  0.155   
4   I   0.330     0.255   0.080        0.2050          0.0895   0.0395  0.055   

   n_rings  
0       15  
1        7  
2        9  
3       10  
4        7

Which this data we are taking a trip under the sea...

The data is all about abalones, which are very large sea snails.

<center><img src="https://upload.wikimedia.org/wikipedia/commons/b/bc/Abalone_at_California_Academy_of_Sciences.JPG" height="500"/></center>

The data is derived from UCI Machine Learning Repository for open data sets. You can find [more information](https://archive.ics.uci.edu/ml/datasets/Abalone) at the UCI page.

The original goal is the find out the age of those sea snails through other easily obtainable physical measurements, without having to cutting them off and count the number of rings, which is pretty time-consuming.
Below you see a table containing all attributes of the snails that have been examined. You found those in as column names in our newly created DataFrame as well.


**Column Description:**

Name |  Data Type | Measurement Unit | Description
:-----|--------|--------|:--------
Sex | nominal | -- | M, F, and I (infant)
Length | continuous | mm | Longest shell measurement
Diameter | continuous | mm | perpendicular to length
Height | continuous | mm | with meat in shell
Whole weight | continuous | grams | whole abalone
Shucked weight | continuous | grams | weight of meat
Viscera weight | continuous | grams | gut weight (after bleeding)
Shell weight | continuous | grams | after being dried
Rings | integer | -- | +1.5 gives the age in years

We already saw the first rows of our DataFrame, when we executed the command `df.head()`. By default this command shows us 5 rows, but we can alter it to any number of rows we want to see:

In [15]:
df.head(8) # or n=8

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010  0.150   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485  0.070   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415  0.210   
3   M   0.440     0.365   0.125        0.5160          0.2155   0.1140  0.155   
4   I   0.330     0.255   0.080        0.2050          0.0895   0.0395  0.055   
5   I   0.425     0.300   0.095        0.3515          0.1410   0.0775  0.120   
6   F   0.530     0.415   0.150        0.7775          0.2370   0.1415  0.330   
7   F   0.545     0.425   0.125        0.7680          0.2940   0.1495  0.260   

   n_rings  
0       15  
1        7  
2        9  
3       10  
4        7  
5        8  
6       20  
7       16

Also, we can check out the last rows in our DataFrame:

In [16]:
df.tail()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
4172   F   0.565     0.450   0.165        0.8870          0.3700   0.2390   
4173   M   0.590     0.440   0.135        0.9660          0.4390   0.2145   
4174   M   0.600     0.475   0.205        1.1760          0.5255   0.2875   
4175   F   0.625     0.485   0.150        1.0945          0.5310   0.2610   
4176   M   0.710     0.555   0.195        1.9485          0.9455   0.3765   

       shell  n_rings  
4172  0.2490       11  
4173  0.2605       10  
4174  0.3080        9  
4175  0.2960       10  
4176  0.4950       12

Two attributes of our DataFrame are highly useful for us to look at:
`shape` and `columns`.

The `shape` will return us the number of rows and columns of our DataFrame

In [17]:
df.shape

(4177, 9)

In [18]:
# gives us back a list of all of the column names
df.columns

Index(['sex', 'length', 'diameter', 'height', 'weight_whole', 'weight_shucked',
       'viscera', 'shell', 'n_rings'],
      dtype='object')

Now we know that we have 4177 individual observations (in this case individual sea snails). In terms of machine learning, we have 8 different features to predict the desired response (output), which in this case are the `n_rings` used to evaluate the age of the sea snails.

In the column description above, the data types of each column is stated. To check whether the data was loaded correctly and the right data types were detected, we can have a look which data types were connected to the columns:

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sex             4177 non-null   object 
 1   length          4177 non-null   float64
 2   diameter        4177 non-null   float64
 3   height          4177 non-null   float64
 4   weight_whole    4177 non-null   float64
 5   weight_shucked  4177 non-null   float64
 6   viscera         4177 non-null   float64
 7   shell           4177 non-null   float64
 8   n_rings         4177 non-null   int64  
dtypes: float64(7), int64(1), object(1)
memory usage: 293.8+ KB


We see that the data was loaded in correctly. To get a feeling of the data, it is highly recommended to get some statistics (better even, a summary of them). This helps us to evaluate the spread and central tendency of out data.

Of course also for this task, there is a pretty neat function already implemented in pandas:


In [20]:
df.describe()

length     diameter       height  weight_whole  weight_shucked  \
count  4177.000000  4177.000000  4177.000000   4177.000000     4177.000000   
mean      0.523992     0.407881     0.139516      0.828742        0.359367   
std       0.120093     0.099240     0.041827      0.490389        0.221963   
min       0.075000     0.055000     0.000000      0.002000        0.001000   
25%       0.450000     0.350000     0.115000      0.441500        0.186000   
50%       0.545000     0.425000     0.140000      0.799500        0.336000   
75%       0.615000     0.480000     0.165000      1.153000        0.502000   
max       0.815000     0.650000     1.130000      2.825500        1.488000   

           viscera        shell      n_rings  
count  4177.000000  4177.000000  4177.000000  
mean      0.180594     0.238831     9.933684  
std       0.109614     0.139203     3.224169  
min       0.000500     0.001500     1.000000  
25%       0.093500     0.130000     8.000000  
50%       0.171000     0.234000     9.000000  
75%       0.253000     0.329000    11.000000  
max       0.760000     1.005000    29.000000

You see, that in this summary, the column `sex`is missing. `.describe()`will only show us numeric columns.
The first row shows us how many observations are available for each feature. The same information was also stored at `.info()`. Luckily in our case, we have no missing data, as the count equals the the # of rows in our DataFrame. This would be different if we would have any `nan` in our DataFrame.

The other values: Mean, Standard deviation, min, max and the quartiles should be familiar to all of you already.

### Grabbing your data

#### The Basics


We now know how to look at our data. What if we wanted to grab certain parts to look at, or certain parts to play around with/transform? Say we wanted to grab an entire row, or an entire column... how do we do that? Let's dive in by starting off with some indexing. 

The format we use to index into our DataFrame and grab data will depend on exactly what subset of the data that we want to grab. If we want to grab entire rows or columns, then we can use bracket notation to do that (just like we use bracket notation to index into lists). If we want an entire column, then we're going to place the **column name** in brackets (and multiple column names in a list inside those brackets). We can also sometimes access a column via dot notation on the DataFrame, which we'll show in a second. If we want an entire row, then we have to place **both** a **beginning and ending index** inside the brackets (it won't work to just place a single index in the brackets). 

In [23]:
# Grabs the "sex" column: bracket notation
df['sex']

0       M
1       M
2       F
3       M
4       I
       ..
4172    F
4173    M
4174    M
4175    F
4176    M
Name: sex, Length: 4177, dtype: object

In [24]:
# Grabs the "sex" column: dot notation
df.sex

0       M
1       M
2       F
3       M
4       I
       ..
4172    F
4173    M
4174    M
4175    F
4176    M
Name: sex, Length: 4177, dtype: object

>Note: The dot notation won't work if the column names contain spaces! That's why it is convenient to check first, if column names contain them and change them into "_" (underscored).

Luckily, the columns in our DataFrame have already underscores.
If not, here is a short code snippet, which you could use to change the column names:

In [25]:
# make a copy of your DataFrame
df2 = df.copy()
# make a list column names
cols = df2.columns.tolist()
# replace space with _
cols = [col.replace(' ', '_') for col in cols]
# reassign new column names to DataFrame
df2.columns = cols

We can also access multiple columns by passing in a list of column names

In [28]:
df[['sex', 'length']]

sex  length
0      M   0.455
1      M   0.350
2      F   0.530
3      M   0.440
4      I   0.330
...   ..     ...
4172   F   0.565
4173   M   0.590
4174   M   0.600
4175   F   0.625
4176   M   0.710

[4177 rows x 2 columns]

Indexing rows:

In [29]:
# This will grab from the beginning up to but not including the row at index 3. 
df[:3]

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010   0.15   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485   0.07   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415   0.21   

   n_rings  
0       15  
1        7  
2        9

In [ ]:
# This will not work because we didn't give it a starting **and** ending index.
df[0]

In [ ]:
# This won't work because we are trying to access a subset of rows 
# **and** columns at the same time. 
df[:1, 'volatile acidity'] 

What if we want to grab certain rows **and** certain columns, rather than just entire rows or entire columns?

If we want to grab only certain rows and columns, there two **methods** that we can use to index into a pandas DataFrame: `loc[]` and `iloc[]`. Note that these are **methods**, which means that we will call them via dot notation on our `DataFrame` object. The difference between these three has to do with how we use them. `loc[]` is a purely label-location based indexer, `iloc[]` is a purely integer-location based indexer.



In [30]:
# Loc is label based. All of these will work, because they are recognized as labels on the 
# rows (index labels) or columns (column name labels). 
df.loc[0, 'sex'] # 0 is one of the index labels, and 'sex' is a column label.


'M'

In [31]:
# Ranges on our index labels still work (as long as they're numeric).
df.loc[0:10, 'sex']

0     M
1     M
2     F
3     M
4     I
5     I
6     F
7     F
8     M
9     F
10    F
Name: sex, dtype: object

In [32]:
df.loc[10:15, ['sex', 'length']]

sex  length
10   F   0.525
11   M   0.430
12   M   0.490
13   F   0.535
14   F   0.470
15   M   0.500

In [ ]:
# These will all fail, because they attempt to access the columns by position integers, 
# and loc only takes labels. 
df.loc[0, 0]
df.loc[0:10, 0]
df.loc[10:15, [0, 4]]

In [33]:
# The above will all work with .iloc, though, since it takes integers (and not labels)
df.iloc[0, 0]
df.iloc[0:10, 0]
df.iloc[10:15, [0, 4]]

sex  weight_whole
10   F        0.6065
11   M        0.4060
12   M        0.5415
13   F        0.6845
14   F        0.4755

In [ ]:
# Using labels, though, like we did with .loc, will NOT work.
df.iloc[0, 'sex']


#### A little bit more

We know how to grab certain rows or columns from DataFrames, as well as a subset of rows and columns, and anything in between. But, this looks like it typically requires that the exact location of the data we want is known. What if we don't know that location? Is there a way to grab desired data by simply specifying some query parameters? Yes! 

There are a couple of ways that we can do this. The first way we'll look at is just through masking, whereas the second actually uses the `query()` method available on the pandas DataFrame.

In [34]:
# This just gives us a mask - tells us True or False whether each row fit's the condition.
df['length'] <= 0.4

0       False
1        True
2       False
3       False
4        True
        ...  
4172    False
4173    False
4174    False
4175    False
4176    False
Name: length, Length: 4177, dtype: bool

In [35]:
# To use a mask, we actually have to use it to index into the DataFrame (using square brackets). 
df[df['length'] <= 0.4]

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
1      M   0.350     0.265   0.090        0.2255          0.0995   0.0485   
4      I   0.330     0.255   0.080        0.2050          0.0895   0.0395   
16     I   0.355     0.280   0.085        0.2905          0.0950   0.0395   
18     M   0.365     0.295   0.080        0.2555          0.0970   0.0430   
20     M   0.355     0.280   0.095        0.2455          0.0955   0.0620   
...   ..     ...       ...     ...           ...             ...      ...   
4151   I   0.350     0.250   0.075        0.1695          0.0835   0.0355   
4152   I   0.370     0.280   0.090        0.2180          0.0995   0.0545   
4162   M   0.385     0.255   0.100        0.3175          0.1370   0.0680   
4163   I   0.390     0.310   0.085        0.3440          0.1810   0.0695   
4164   I   0.390     0.290   0.100        0.2845          0.1255   0.0635   

       shell  n_rings  
1     0.0700        7  
4     0.0550        7  
16    0.1150        7  
18    0.1000        7  
20    0.0750       11  
...      ...      ...  
4151  0.0410        6  
4152  0.0615        7  
4162  0.0920        8  
4163  0.0790        7  
4164  0.0810        7  

[695 rows x 9 columns]

Notice how only the indices that were found to be True from the condition show up in this subset of the DataFrame. We've "masked" off the rest of the indices that we're found to be False (hence the name **masking**). 

In [37]:
# Okay, this is cool. What if I wanted a slightly more complicated query...
df[(df['length'] <= 0.4) & (df['sex'] == 'F')].head()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
71    F   0.400     0.320   0.110        0.3530          0.1405   0.0985   
132   F   0.325     0.260   0.090        0.1915          0.0850   0.0360   
136   F   0.305     0.230   0.080        0.1560          0.0675   0.0345   
226   F   0.390     0.290   0.125        0.3055          0.1210   0.0820   
331   F   0.400     0.325   0.120        0.3185          0.1340   0.0565   

     shell  n_rings  
71   0.100        8  
132  0.062        7  
136  0.048        7  
226  0.090        7  
331  0.095        8

In [38]:
# So I could write an arbitrarily complicated query using that syntax... 
df[(df['length'] <= 0.4) & (df['length'] > 0.35) & (df['weight_whole'] > 0.25) & (df['weight_whole'] < 0.3)]

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
16     I   0.355     0.280   0.085        0.2905          0.0950   0.0395   
18     M   0.365     0.295   0.080        0.2555          0.0970   0.0430   
121    I   0.385     0.295   0.085        0.2535          0.1030   0.0575   
127    I   0.385     0.290   0.085        0.2505          0.1120   0.0610   
139    M   0.375     0.285   0.095        0.2530          0.0960   0.0575   
...   ..     ...       ...     ...           ...             ...      ...   
3923   M   0.395     0.280   0.080        0.2660          0.0995   0.0660   
3968   I   0.375     0.290   0.095        0.2875          0.1230   0.0605   
3969   I   0.380     0.300   0.090        0.2770          0.1655   0.0625   
3971   I   0.400     0.295   0.095        0.2520          0.1105   0.0575   
4164   I   0.390     0.290   0.100        0.2845          0.1255   0.0635   

       shell  n_rings  
16    0.1150        7  
18    0.1000        7  
121   0.0850        7  
127   0.0800        8  
139   0.0925        9  
...      ...      ...  
3923  0.0900       12  
3968  0.0800        6  
3969  0.0820        6  
3971  0.0660        6  
4164  0.0810        7  

[86 rows x 9 columns]

In [39]:
# Or I could use the query() method that is available on our DataFrame object. 
df.query('length <= 0.4 and length > 0.35 and weight_whole > 0.25 and weight_whole < 0.3')

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
16     I   0.355     0.280   0.085        0.2905          0.0950   0.0395   
18     M   0.365     0.295   0.080        0.2555          0.0970   0.0430   
121    I   0.385     0.295   0.085        0.2535          0.1030   0.0575   
127    I   0.385     0.290   0.085        0.2505          0.1120   0.0610   
139    M   0.375     0.285   0.095        0.2530          0.0960   0.0575   
...   ..     ...       ...     ...           ...             ...      ...   
3923   M   0.395     0.280   0.080        0.2660          0.0995   0.0660   
3968   I   0.375     0.290   0.095        0.2875          0.1230   0.0605   
3969   I   0.380     0.300   0.090        0.2770          0.1655   0.0625   
3971   I   0.400     0.295   0.095        0.2520          0.1105   0.0575   
4164   I   0.390     0.290   0.100        0.2845          0.1255   0.0635   

       shell  n_rings  
16    0.1150        7  
18    0.1000        7  
121   0.0850        7  
127   0.0800        8  
139   0.0925        9  
...      ...      ...  
3923  0.0900       12  
3968  0.0800        6  
3969  0.0820        6  
3971  0.0660        6  
4164  0.0810        7  

[86 rows x 9 columns]

In general, it's preferred to use the `query()` method, since it improves readability. It doesn't use loads of sets of brackets (`[]`) and parentheses (`()`), but rather just one set of parentheses. It also tends to follow the Python syntax a little more closely than the mask methods that we looked at above, using `and` instead of `&` to separate different specifications on our queries. 


We're going to look at a couple of more things you can do with DataFrames, but to view all available attributes and methods of DataFrames, we can check out the [pandas Docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html). For practical examples of how DataFrames are used you can get a copy of [Python for Data Analysis](http://shop.oreilly.com/product/0636920023784.do) (it's written by Wes McKinney, the creator of pandas).

For those coming from R, know that the pandas DataFrame was based off the R DataFrame, and most anything we can do with an R DataFrame, we can do with a pandas DataFrame. For anybody coming from a SQL background, the methods available via DataFrame's give us much, if not all, of the functionality that we have available in your SQL environment. 


#### Groupby

Let's start with `groupby`s...

By “group by” we are referring to a process involving one or more of the following steps

- __Splitting__ the data into groups based on some criteria
- __Applying__ a function to each group independently
- __Combining__ the results into a data structure

[source](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html)

In [40]:
# first we try to groupby the sex, let's check how many different values this feature can have 
df['sex'].unique()

array(['M', 'F', 'I'], dtype=object)

In [41]:
# we can also see the count of different values with this command
df['sex'].nunique()

3

In [42]:
df.groupby('sex')   # Note that this returns back to us a groupby object. It doesn't actually 
                    # return to us anything useful until we perform some aggregation on it. 

We have tons of aggregation metrics we can get from a groupby object. Note here that we store the results of a groupby below to then perform all kinds of operations on it (this is actually the preferred method if we're going to perform more than one calculation on it).

In [43]:
# Group by sex column
groupby_obj = df.groupby('sex')

# Performing an aggregation by calculating the mean
groupby_obj.mean()

length  diameter    height  weight_whole  weight_shucked   viscera  \
sex                                                                         
F    0.579093  0.454732  0.158011      1.046532        0.446188  0.230689   
I    0.427746  0.326494  0.107996      0.431363        0.191035  0.092010   
M    0.561391  0.439287  0.151381      0.991459        0.432946  0.215545   

        shell    n_rings  
sex                       
F    0.302010  11.129304  
I    0.128182   7.890462  
M    0.281969  10.705497

In [44]:
groupby_obj.max()


length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
sex                                                                           
F     0.815      0.65   1.130        2.6570          1.4880   0.5900  1.005   
I     0.725      0.55   0.220        2.0495          0.7735   0.4405  0.655   
M     0.780      0.63   0.515        2.8255          1.3510   0.7600  0.897   

     n_rings  
sex           
F         29  
I         21  
M         27

In [45]:
groupby_obj.count()

length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
sex                                                                           
F      1307      1307    1307          1307            1307     1307   1307   
I      1342      1342    1342          1342            1342     1342   1342   
M      1528      1528    1528          1528            1528     1528   1528   

     n_rings  
sex           
F       1307  
I       1342  
M       1528

The previous aggregation metrics gave us back a DataFrame with all of the columns minus what we grouped on. Notice that what we grouped on becomes the index. What if I wanted only one column back (especially with something like count, where it is the same for every column)? 
Well, we can do anything with this DataFrame that we did before...

In [46]:
df.groupby('sex').count()['length']

sex
F    1307
I    1342
M    1528
Name: length, dtype: int64

Note we can also group by multiple columns by passing them in in a list. It will group by the first column passed in first, and then the second after that (i.e. it will group by the second within the group by of the first). 

In [47]:
# Note we can also group by multiple columns by passing them in a list. It will group by 
# the first column passed in first, and then the second after that (i.e. it will group by 
# the second within the group by of the first). 
df.groupby(['sex', 'n_rings']).count()['weight_whole']

sex  n_rings
F    5            4
     6           16
     7           44
     8          122
     9          238
               ... 
M    22           3
     23           3
     24           1
     26           1
     27           1
Name: weight_whole, Length: 68, dtype: int64

Check out the [Group By documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html) to look at what all you can do with the pandas `.groupby()`.

#### Sorting 


Sorting is going to work much the same way as using `groupby`. It is going to be available via a method that we call on the DataFrame, `.sort_values()`, and we are going to pass it a column or columns to sort by. 

In [48]:
df.sort_values('length').head(15) # Note: this is ascending by default.
# df.sort_values('length', ascending=False) # here we change ascending to false.

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
236    I   0.075     0.055   0.010        0.0020          0.0010   0.0005   
238    I   0.110     0.090   0.030        0.0080          0.0025   0.0020   
2114   I   0.130     0.095   0.035        0.0105          0.0050   0.0065   
237    I   0.130     0.100   0.030        0.0130          0.0045   0.0030   
1986   I   0.135     0.130   0.040        0.0290          0.0125   0.0065   
1429   I   0.140     0.105   0.035        0.0140          0.0055   0.0025   
3899   I   0.140     0.105   0.035        0.0145          0.0050   0.0035   
719    I   0.150     0.100   0.025        0.0150          0.0045   0.0040   
2381   M   0.155     0.115   0.025        0.0240          0.0090   0.0050   
696    I   0.155     0.105   0.050        0.0175          0.0050   0.0035   
526    M   0.155     0.110   0.040        0.0155          0.0065   0.0030   
720    I   0.160     0.110   0.025        0.0180          0.0065   0.0055   
239    I   0.160     0.120   0.035        0.0210          0.0075   0.0045   
1987   I   0.160     0.110   0.025        0.0195          0.0075   0.0050   
3902   I   0.160     0.120   0.020        0.0180          0.0075   0.0045   

       shell  n_rings  
236   0.0015        1  
238   0.0030        3  
2114  0.0035        4  
237   0.0040        3  
1986  0.0080        4  
1429  0.0040        3  
3899  0.0050        4  
719   0.0050        2  
2381  0.0075        5  
696   0.0050        4  
526   0.0050        3  
720   0.0050        3  
239   0.0050        5  
1987  0.0060        4  
3902  0.0050        4

We can sort by multiple columns by placing them in a list inside of the sort() method. It will sort by the first column passed in first, and then the second within the sort of the first.

In [49]:
# Sort by multiple columns
df.sort_values(['length', 'diameter'], ascending=False).head(15) # ascending=False will apply to both columns. 

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
1428   F   0.815     0.650   0.250        2.2550          0.8905   0.4200   
2334   F   0.800     0.630   0.195        2.5260          0.9330   0.5900   
1209   F   0.780     0.630   0.215        2.6570          1.4880   0.4985   
3715   M   0.780     0.600   0.210        2.5480          1.1945   0.5745   
1763   M   0.775     0.630   0.250        2.7795          1.3485   0.7600   
2090   M   0.775     0.570   0.220        2.0320          0.7350   0.4755   
1762   M   0.770     0.620   0.195        2.5155          1.1155   0.6415   
4148   M   0.770     0.605   0.175        2.0505          0.8005   0.5260   
2625   M   0.770     0.600   0.215        2.1945          1.0515   0.4820   
1052   M   0.765     0.600   0.220        2.3020          1.0070   0.5090   
2624   M   0.765     0.585   0.180        2.3980          1.1280   0.5120   
1985   M   0.760     0.605   0.215        2.1730          0.8010   0.4915   
3896   M   0.760     0.575   0.190        1.8290          0.7035   0.3860   
1207   F   0.755     0.625   0.210        2.5050          1.1965   0.5130   
1208   M   0.755     0.580   0.205        2.0065          0.8295   0.4015   

       shell  n_rings  
1428  0.7975       14  
2334  0.6200       23  
1209  0.5860       11  
3715  0.6745       11  
1763  0.5780       12  
2090  0.6585       17  
1762  0.6420       12  
4148  0.3550       11  
2625  0.5840       10  
1052  0.6205       12  
2624  0.5335       12  
1985  0.6460       13  
3896  0.5600       14  
1207  0.6785       11  
1208  0.5950       10

#### Creating and Dropping Columns

Creating columns is done in one of two ways: 
1. Using bracket notation
2. Using the `eval()` method on the pandas DataFrame. 

Dropping columns is done using the `df.drop()` method on the pandas DataFrame. When dropping columns, we have to be careful to make sure to tell the DataFrame to drop them in place, or assign the DataFrame with dropped columns to a new variable. You also need to make sure to tell the `drop()` method what axis the thing you're trying to drop is on (rows are `axis=0`, and columns are `axis=1`).
![](https://i.stack.imgur.com/DL0iQ.jpg)

In [50]:
# we can rename columns with `.rename()`
df.rename(columns={'n_rings': 'nr_rings'}, inplace=True)
df.head()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010  0.150   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485  0.070   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415  0.210   
3   M   0.440     0.365   0.125        0.5160          0.2155   0.1140  0.155   
4   I   0.330     0.255   0.080        0.2050          0.0895   0.0395  0.055   

   nr_rings  
0        15  
1         7  
2         9  
3        10  
4         7

In [52]:
# with `.eval()` we can create new columns
# remember that in the description it is stated that the age of the snails can be calculated by adding 1.5 to the nr. of rings
df.eval('age = nr_rings + 1.5', inplace = True)
df.head()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010  0.150   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485  0.070   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415  0.210   
3   M   0.440     0.365   0.125        0.5160          0.2155   0.1140  0.155   
4   I   0.330     0.255   0.080        0.2050          0.0895   0.0395  0.055   

   nr_rings   age  
0        15  16.5  
1         7   8.5  
2         9  10.5  
3        10  11.5  
4         7   8.5

In [53]:
df.eval('weight_per_hight = weight_whole / diameter', inplace = True)
df.columns

Index(['sex', 'length', 'diameter', 'height', 'weight_whole', 'weight_shucked',
       'viscera', 'shell', 'nr_rings', 'age', 'weight_per_hight'],
      dtype='object')

In [54]:
df.head()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010  0.150   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485  0.070   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415  0.210   
3   M   0.440     0.365   0.125        0.5160          0.2155   0.1140  0.155   
4   I   0.330     0.255   0.080        0.2050          0.0895   0.0395  0.055   

   nr_rings   age  weight_per_hight  
0        15  16.5          1.408219  
1         7   8.5          0.850943  
2         9  10.5          1.611905  
3        10  11.5          1.413699  
4         7   8.5          0.803922

Sometimes it is also necessary to delete columns, e.g. if they do not contain relevant data, or if engineered features contain the same information.

The age is our newly created feature, therefore nr_rings are not containing any further information with axis=1 we define that we are dropping a column, not a row by default it would be axis=0.

In [55]:
# the age is our newly created feature, therefore "nr_rings" does not contain any further information
# with axis=1 we define that we are dropping a column, not a row
# by default it would be axis=0
df.drop('nr_rings', axis=1)

sex  length  diameter  height  weight_whole  weight_shucked  viscera  \
0      M   0.455     0.365   0.095        0.5140          0.2245   0.1010   
1      M   0.350     0.265   0.090        0.2255          0.0995   0.0485   
2      F   0.530     0.420   0.135        0.6770          0.2565   0.1415   
3      M   0.440     0.365   0.125        0.5160          0.2155   0.1140   
4      I   0.330     0.255   0.080        0.2050          0.0895   0.0395   
...   ..     ...       ...     ...           ...             ...      ...   
4172   F   0.565     0.450   0.165        0.8870          0.3700   0.2390   
4173   M   0.590     0.440   0.135        0.9660          0.4390   0.2145   
4174   M   0.600     0.475   0.205        1.1760          0.5255   0.2875   
4175   F   0.625     0.485   0.150        1.0945          0.5310   0.2610   
4176   M   0.710     0.555   0.195        1.9485          0.9455   0.3765   

       shell   age  weight_per_hight  
0     0.1500  16.5          1.408219  
1     0.0700   8.5          0.850943  
2     0.2100  10.5          1.611905  
3     0.1550  11.5          1.413699  
4     0.0550   8.5          0.803922  
...      ...   ...               ...  
4172  0.2490  12.5          1.971111  
4173  0.2605  11.5          2.195455  
4174  0.3080  10.5          2.475789  
4175  0.2960  11.5          2.256701  
4176  0.4950  13.5          3.510811  

[4177 rows x 10 columns]

In [ ]:
df.columns # Wait, the nr_rings column is still there... why? 
           # It's because we didn't tell it drop inplace. 

In [ ]:
df.drop('nr_rings', inplace=True, axis=1) 

In [ ]:
df.columns

### Dealing with Nulls

pandas has functions for both filling nulls (or N/As) with whatever value we want, or dropping nulls all together. To fill nulls, we use the `.fillna()` method on the DataFrame, and to drop nulls, we call the `.dropna()` method on the DataFrame. In terms of the `.fillna()` function, we can give it a default value to fill in, or a number of other methods to fill it in (padding, back filling, forward filling). You can read about dealing with missing data in the docs [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html). We're not going to go into too much depth here, but want you to know that this functionality exists. 

As we already checked, in this dataset, there are not even missing data. So the next line of code will not do anything in this case. Just so you know how to fill nans or drop them with pandas.

In [57]:
# Notice the addition of the inplace argument here. 
# if you try to fill nans will some value: the first argument is the inserted value
df.fillna(-1, inplace=True)
# dropna() will drop the whole row, if nans are present
df.dropna(inplace=True) 
df.head()

sex  length  diameter  height  weight_whole  weight_shucked  viscera  shell  \
0   M   0.455     0.365   0.095        0.5140          0.2245   0.1010  0.150   
1   M   0.350     0.265   0.090        0.2255          0.0995   0.0485  0.070   
2   F   0.530     0.420   0.135        0.6770          0.2565   0.1415  0.210   
3   M   0.440     0.365   0.125        0.5160          0.2155   0.1140  0.155   
4   I   0.330     0.255   0.080        0.2050          0.0895   0.0395  0.055   

   nr_rings   age  weight_per_hight  
0        15  16.5          1.408219  
1         7   8.5          0.850943  
2         9  10.5          1.611905  
3        10  11.5          1.413699  
4         7   8.5          0.803922

## A quick aside on Pandas Series

You might have noticed that in a couple of places, when we asked for certain rows/columns of the data, we got back a 1-D array that had an index attached. These are examples of what pandas calls `Series`. In the documentation for [pandas Series](http://pandas.pydata.org/pandas-docs/version/0.15.2/dsintro.html#series), you can get an idea of what they can do. For the most part, we can kind of treat them like a mini DataFrame, as they have a lot of the same methods. However, there are some slight differences. Since we work with DataFrame's the majority of the time, we're not going to go into any real depth on Series. 

Here are some examples of things that returned series: 

In [ ]:
df['length'] <= 0.3

In [ ]:
type(df['length'] <= 0.3)

In [58]:
df.groupby('sex').count()['age']

sex
F    1307
I    1342
M    1528
Name: age, dtype: int64

That was a lot of input! Don't worry, you can test and solidify your knowledge in the next notebooks.

---

## Summary

- pandas' DataFrame is the de facto data structure for Data Scientists
- Get your data into a DataFrames ASAP
- Use built-in methods to go fast!
    - For example, use groupby to calculate aggregate statistics for categories

----
## Further Study

There are many more functions on DataFrame you can discover. Here we have put together a few more links that describe the functionality of other functions and have great code snippets that you can use in your next projects.

- [Helpful Python Code Snippets for Data Exploration in Pandas](https://medium.com/@msalmon00/helpful-python-code-snippets-for-data-exploration-in-pandas-b7c5aed5ecb9)
- [Manipulating tabular data with Pandas](https://neuroimaging-data-science.org/content/004-scipy/002-pandas.html)
- [Book - Python for Data Analysis](http://shop.oreilly.com/product/0636920023784.do)